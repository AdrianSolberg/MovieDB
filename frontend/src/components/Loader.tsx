import { cn } from "@/shadcn/lib/utils";
import React from "react";

type LoaderProps = {
    color?: "stroke-primary" | "stroke-secondary";
    size?: "sm" | "md" | "lg";
} & React.HTMLAttributes<HTMLDivElement>;

const Loader = React.forwardRef<
    HTMLDivElement,
    React.PropsWithChildren<LoaderProps>
>(
    (
        {
            className,
            children,
            color = "stroke-primary",
            size = "md",
            ...props
        },
        ref
    ) => {
        const sizeClass =
            size === "sm"
                ? "h-6 w-6"
                : size === "lg"
                  ? "h-16 w-16"
                  : "h-12 w-12";

        return (
            <div
                ref={ref}
                className={cn(
                    "m-2 flex flex-col items-center justify-center gap-2",
                    className
                )}
                role="status"
                aria-live="polite"
                {...props}
            >
                <svg
                    viewBox="0 0 800 800"
                    xmlns="http://www.w3.org/2000/svg"
                    className={cn(
                        "animate-spin stroke-primary",
                        sizeClass,
                        color
                    )}
                >
                    <circle
                        cx="400"
                        cy="400"
                        fill="none"
                        r="350"
                        strokeWidth="64"
                        strokeDasharray="1648,550"
                        strokeLinecap="round"
                    />
                </svg>
                {children ? (
                    children
                ) : (
                    <span className="sr-only">Loading...</span>
                )}
            </div>
        );
    }
);

export default Loader;
