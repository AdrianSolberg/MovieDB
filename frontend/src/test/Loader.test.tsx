import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import Loader from "@/components/Loader";
import React from "react";

describe("Loader", () => {
    it("matches snapshot", () => {
        const { asFragment } = render(<Loader />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders with default props", () => {
        render(<Loader />);

        const loaderDiv = screen.getByRole("status");
        expect(loaderDiv).toBeInTheDocument();
        expect(loaderDiv).toHaveAttribute("aria-live", "polite");

        const loaderSvg = screen.getByRole("status").querySelector("svg");
        expect(loaderSvg).toHaveClass("h-12 w-12");
        expect(loaderSvg).toHaveClass("stroke-primary");

        const srOnly = screen.getByText("Loading...");
        expect(srOnly).toHaveClass("sr-only");
    });

    it("renders with different size", () => {
        render(<Loader size="sm" />);
        const loaderSvg = screen.getByRole("status").querySelector("svg");
        expect(loaderSvg).toHaveClass("h-6 w-6");
    });

    it("renders with large size", () => {
        render(<Loader size="lg" />);
        const loaderSvg = screen.getByRole("status").querySelector("svg");
        expect(loaderSvg).toHaveClass("h-16 w-16");
    });

    it("renders with secondary color", () => {
        render(<Loader color="stroke-secondary" />);
        const loaderSvg = screen.getByRole("status").querySelector("svg");
        expect(loaderSvg).toHaveClass("stroke-secondary");
    });

    it("renders with additional children", () => {
        render(<Loader>Custom Loading Text</Loader>);
        const customText = screen.getByText("Custom Loading Text");
        expect(customText).toBeInTheDocument();

        const srOnly = screen.queryByText("Loading...");
        expect(srOnly).not.toBeInTheDocument();
    });

    it("applies additional className", () => {
        render(<Loader className="custom-class" />);
        const loaderDiv = screen.getByRole("status");
        expect(loaderDiv).toHaveClass("custom-class");
    });

    it("forwards ref", () => {
        const ref = React.createRef<HTMLDivElement>();
        render(<Loader ref={ref} />);
        expect(ref.current).toBeTruthy();
    });
});
