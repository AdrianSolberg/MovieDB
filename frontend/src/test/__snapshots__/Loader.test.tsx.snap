// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loader > matches snapshot 1`] = `
<DocumentFragment>
  <div
    aria-live="polite"
    class="m-2 flex flex-col items-center justify-center gap-2"
    role="status"
  >
    <svg
      class="animate-spin h-12 w-12 stroke-primary"
      viewBox="0 0 800 800"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="400"
        cy="400"
        fill="none"
        r="350"
        stroke-dasharray="1648,550"
        stroke-linecap="round"
        stroke-width="64"
      />
    </svg>
    <span
      class="sr-only"
    >
      Loading...
    </span>
  </div>
</DocumentFragment>
`;
